File: python/day17.py
import itertools


def count_neighbors(field, pos):
    if len(field) == 0:
        return 0
    res = 0
    for p in itertools.product(*[range(x - 1, x + 2) for x in pos]):
        if p != pos and p in field:
            res += 1
    return res


def get_bounds(field):
    if len(field) == 0:
        return None
    val = next(iter(field))
    res = [[x, x] for x in val]
    for pos in field:
        for i, x in enumerate(pos):
            res[i][0] = min(res[i][0], x)
            res[i][1] = max(res[i][1], x)
    return res


def step(field):
    bounds = get_bounds(field)
    res = set()
    for pos in itertools.product(*[range(b[0] - 1, b[1] + 2) for b in bounds]):
        n = count_neighbors(field, pos)
        if pos in field:
            if n == 2 or n == 3:
                res.add(pos)
        else:
            if n == 3:
                res.add(pos)
    return res


def init_field(seed, ndim=3):
    field = set()
    for y in range(len(seed)):
        for x in range(len(seed[y])):
            if seed[y][x] == '#':
                field.add(tuple([x, y] + [0] * (ndim - 2)))
    return field


NUM_STEPS = 6


def run(seed, ndim, num_steps=NUM_STEPS):
    field = init_field(seed, ndim)
    for i in range(num_steps):
        field = step(field)
    return len(field)


def solution():
    seed = [line.strip() for line in open('../data/17.txt', 'r').readlines()]
    print(f'Answer 1: {run(seed, 3)}')
    print(f'Answer 2: {run(seed, 4)}')

File: rust/src/day17.rs
use crate::common;
use std::collections::HashSet;

extern crate itertools;
use itertools::Itertools;

type Field = HashSet<Vec<i16>>;

fn count_neighbors(field: &Field, pos: &Vec<i16>) -> usize {
  pos
    .iter()
    .map(|x| (x - 1)..=(x + 1))
    .multi_cartesian_product()
    .filter(|p| p != pos && field.contains(p))
    .count()
}

fn get_bounds(field: &Field) -> Option<Vec<(i16, i16)>> {
  if field.len() == 0 {
    return None;
  }
  let pos0 = field.iter().next().unwrap();
  let mut res: Vec<(i16, i16)> = pos0.iter().map(|x| (*x, *x)).collect();
  for pos in field.iter() {
    for (i, x) in pos.iter().enumerate() {
      res[i] = (*x.min(&res[i].0), *x.max(&res[i].1));
    }
  }
  Some(res)
}

fn step(field: &Field) -> Field {
  let mut res = Field::new();
  let bounds = get_bounds(&field);
  if bounds == None {
    return res;
  }
  let coord_it = bounds
    .unwrap()
    .iter()
    .map(|(a, b)| (a - 1)..=(b + 1))
    .multi_cartesian_product();
  for pos in coord_it {
    let n = count_neighbors(field, &pos);
    if field.contains(&pos) {
      if n == 2 || n == 3 {
        res.insert(pos);
      }
    } else {
      if n == 3 {
        res.insert(pos);
      }
    }
  }
  res
}

fn init_field(seed: &Vec<String>, ndim: usize) -> Field {
  let mut res = Field::new();
  for (y, line) in seed.iter().enumerate() {
    for (x, c) in line.as_bytes().iter().enumerate() {
      if *c as char == '#' {
        let mut key: Vec<i16> = Vec::new();
        key.resize(ndim, 0);
        key[0] = x as i16;
        key[1] = y as i16;
        res.insert(key);
      }
    }
  }
  res
}

const NUM_ITER: usize = 6;

fn run(seed: &Vec<String>, ndim: usize) -> usize {
  let mut field = init_field(&seed, ndim);
  for _ in 0..NUM_ITER {
    field = step(&field);
  }
  field.len()
}

pub(crate) fn solution() {
  let seed = common::read_lines(&common::data_file(17)).unwrap();
  println!("Answer 1: {}", run(&seed, 3));
  println!("Answer 2: {}", run(&seed, 4));
}
